cri: Add background stats collector to calculate UsageNanoCores

dims opened the pull request:
adds a background stats collector that calculates `UsageNanoCores` for containers and pod sandboxes.

- run in the background every second to collect CPU metrics for all containers and sandboxes (similar to what cAdvisor does)
- keep a rolling buffer of CPU samples and calculates the instantaneous CPU usage rate from consecutive samples
- read pod-level CPU stats from the parent cgroup rather than the pause container
- add cgroupv2 Pressure Stall Information for CPU, memory, and IO
- add missing `Timestamp` and `Interfaces` fields

when Kubernetes runs with `PodAndContainerStatsFromCRI=true`, it expects `UsageNanoCores` to be set in stats responses. This value represents how much CPU is being used right now (as opposed to `UsageCoreNanoSeconds` which is cumulative). To calculate it, we need to compare CPU samples over time to replicate what is in cadvisor.

we can't yet really test this in CI as some changes in kubernetes has to land for `--feature-gates=PodAndContainerStatsFromCRI=true` 

PS: see https://github.com/containerd/containerd/pull/12620 where we tested against a forked branch to get things working

```release-note
Add background stats collector to calculate UsageNanoCores for containers and pod sandboxes
```

dims committed:
cri: Add background stats collector to calculate UsageNanoCores

adds a background stats collector that calculates `UsageNanoCores` for containers and pod sandboxes.

- run in the background every second to collect CPU metrics for all containers and sandboxes (similar to what cAdvisor does)
- keep a rolling buffer of CPU samples and calculates the instantaneous CPU usage rate from consecutive samples
- read pod-level CPU stats from the parent cgroup rather than the pause container
- add cgroupv2 Pressure Stall Information for CPU, memory, and IO
- add missing `Timestamp` and `Interfaces` fields

when Kubernetes runs with `PodAndContainerStatsFromCRI=true`, it expects `UsageNanoCores` to be set in stats responses.
This value represents how much CPU is being used right now (as opposed to `UsageCoreNanoSeconds` which is cumulative).
To calculate it, we need to compare CPU samples over time to replicate what is in cadvisor.

we can't yet really test this in CI as some changes in kubernetes has to land for `--feature-gates=PodAndContainerStatsFromCRI=true`

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims commented:
cc @mikebrow @akhilerm

dims commented:
xref: https://github.com/kubernetes/kubernetes/pull/135604

dims commented:
@akhilerm @mikebrow please review when you get a chance. I have not yet turned on the feature flags to allow this to be tested as that needs https://github.com/kubernetes/kubernetes/pull/135604 to land first after k8s code freeze. Once you are happy we can land this one to prepare for it (chicken/egg!)

mikebrow commented on internal/cri/server/container_stats_list.go:
```diff
@@ -179,6 +179,17 @@ func (c *criService) toCRIContainerStats(css []containerStats) *runtime.ListCont
+	// First, try to get pre-calculated UsageNanoCores from the background stats collector.
+	// This ensures we have valid data even on the first query (as the collector runs
+	// continuously in the background, similar to cAdvisor's housekeeping).
+	if c.statsCollector != nil {
```
or platform is other...

mikebrow commented on internal/cri/server/stats_collector.go:
```diff
@@ -0,0 +1,300 @@
+	defaultCollectionInterval = 1 * time.Second
+
+	// defaultStatsAge is how long stats samples are retained.
+	defaultStatsAge = 2 * time.Minute
```
should probably calculate this one

dims replied:
you mean calculate `defaultMaxSamples` right?

dims replied:
Done!

mikebrow commented on internal/cri/server/stats_collector.go:
```diff
@@ -0,0 +1,300 @@
+const (
+	// defaultCollectionInterval is how often the collector fetches metrics.
+	// This matches cAdvisor's default housekeeping interval.
+	defaultCollectionInterval = 1 * time.Second
```
maybe these should be tunable values, 120 1sec intervals / each container seems arbitrary and expensive

dims replied:
2 mins comes from here - https://github.com/google/cadvisor/blob/b21c1a151105736b8307e8ae88bde5e9a278ac7f/cmd/storagedriver.go#L38

1 second comes from here - https://github.com/google/cadvisor/blob/b21c1a151105736b8307e8ae88bde5e9a278ac7f/manager/container.go#L50

(trying to match what we have currently!)

dims replied:
whoops! on it

mikebrow commented on internal/cri/server/container_stats_list.go:
```diff
@@ -179,6 +179,17 @@ func (c *criService) toCRIContainerStats(css []containerStats) *runtime.ListCont
+	// First, try to get pre-calculated UsageNanoCores from the background stats collector.
+	// This ensures we have valid data even on the first query (as the collector runs
+	// continuously in the background, similar to cAdvisor's housekeeping).
+	if c.statsCollector != nil {
```
should we check platform here given container stats collector is not implemented outside linux

mikebrow replied:
nod but with the original old model that we'll drop down to below that still runs for the other platforms .. (but not for sandbox as sandbox is no-op for other platforms in the older model...

mikebrow commented on internal/cri/server/stats_collector.go:
```diff
@@ -0,0 +1,300 @@
+
+// RemoveContainer removes the stats store for a container/sandbox.
+// Should be called when a container is removed.
+func (c *StatsCollector) RemoveContainer(id string) {
```
this func is not being called yet when the container is removed from the container store..

mikebrow reviewed (commented):
Very nice!  see comments..

once we have "switched" to timed for a container.. do we need to/should we flush the original old method ContainerStats by setting it back to nil

this might be pretty expensive to collect and stage. there is a risk that ticker gets backed up due to locks in container store list and callouts to task metrics for the set of all k8s.io container ids....

mikebrow commented on internal/cri/server/service.go:
```diff
@@ -265,6 +267,11 @@ func (c *criService) Run(ready func()) error {
+	// Start the background stats collector for UsageNanoCores calculation
+	log.L.Info("Start stats collector")
+	c.statsCollector = NewStatsCollector(c)
+	c.statsCollector.Start()
```
check for nil here...

mikebrow commented on internal/cri/server/container_remove.go:
```diff
@@ -59,6 +59,9 @@ func (c *criService) RemoveContainer(ctx context.Context, r *runtime.RemoveConta
 		// from the container store as well and return successfully.
 		log.G(ctx).WithError(err).Warn("get container info failed")
 		c.containerStore.Delete(ctrID)
+		if c.statsCollector != nil {
```
over here I think.. https://github.com/containerd/containerd/blob/main/internal/cri/store/container/container.go#L197

mikebrow replied:
yeah but it's tied to the life of the container id... which are reusable ids

dims replied:
i am missing something in `func (s *Store) Delete(id string)` i don't have access to either `criService` or `statsCollector` member of `criService`...

mikebrow commented on internal/cri/server/container_create.go:
```diff
@@ -462,6 +462,9 @@ func (c *criService) createContainer(r *createContainerRequest) (_ string, retEr
 	if err := c.containerStore.Add(container); err != nil {
 		return "", fmt.Errorf("failed to add container %q into store: %w", r.containerID, err)
 	}
+	if c.statsCollector != nil {
```
same here.. over in the container store is best given some actions, restart etc.. don't involve a CRI call

dims replied:
Sorry i lost you, what should i do here?

dims replied:
ack!

dims replied:
@mikebrow hmm now i remember, c.statsCollector is not available there in the Create method and i did not want to pass that along!

mikebrow commented on internal/cri/config/config.go:
```diff
@@ -427,6 +427,19 @@ type RuntimeConfig struct {
 	// that should be ignored for checking "ContainerdHasNoDeprecationWarnings" condition.
 	IgnoreDeprecationWarnings []string `toml:"ignore_deprecation_warnings" json:"ignoreDeprecationWarnings"`
+
+	// StatsCollectionPeriod is the period for collecting container/sandbox CPU stats
```
docs/cri/config.md :-)

dims replied:
Dang! on it.

mikebrow reviewed (approved):
LGTM
~~needs rebase to pickup fix for the fuzzer setup~~

akhilerm commented on internal/cri/server/sandbox_stats_linux.go:
```diff
@@ -76,14 +83,17 @@ func (c *criService) podSandboxStats(
-			},
+			Timestamp:        timestamp.UnixNano(),
+			DefaultInterface: defaultInterface,
+			Interfaces:       []*runtime.NetworkInterfaceUsage{defaultInterface},
```
This should not contain the network usage from the default interface. Ref: https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1/api.proto#L765-L766

akhilerm commented on internal/cri/server/stats_collector.go:
```diff
@@ -0,0 +1,335 @@
+	// stopOnce ensures Stop() only closes stopCh once
+	stopOnce sync.Once
+	// service provides access to task metrics
+	service *criService
```
`StatsCollector` is already part of `criService` struct https://github.com/containerd/containerd/pull/12629/files#diff-3bbef1ddd83b801d65af1f7d796e5193fcf66158aa7d554a22292c4e327a38a5R164. So why have the service again as a field here?

akhilerm reviewed (commented):
Got a few questions and comments.

dims committed:
stats_collection_period -> stats_collect_period

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Removed the circular dependency

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
cri: simplify network stats to only add Timestamp field

Remove unnecessary variable extraction and Interfaces field,
keeping only the Timestamp addition as originally intended.

Signed-off-by: Davanum Srinivas <davanum@gmail.com>
